src/io/flutter/preview/render_server_template.txt

```diff
@@ -0,0 +1,192 @@
+    }
+  }
+
+  static ElementInfo buildHierarchy(
+      Map<Widget, Element> elementsMap, Map<int, Widget> registeredWidgets) {
+    var elementToId = new Map<Element, int>.identity();
+    var elementToInfoMap = <Element, ElementInfo>{};
+
+    registeredWidgets.forEach((id, widget) {
+      final Element element = elementsMap[widget];
+      assert(element != null);
+      elementToId[element] = id;
+    });
+
+    Element getInterestingAncestor(Element element) {
+      Element ancestor;
+      element.visitAncestorElements((Element candidate) {
+        ancestor = candidate;
+        return !elementToId.containsKey(candidate);
+      });
+      return ancestor;
+    }
+
+    ElementInfo createInfoForElement(Element element) {
+      var id = elementToId[element];
+      if (id == null) {
+        return null;
+      }
+
+      var info = elementToInfoMap[element];
+      if (info == null) {
+        Element ancestor = getInterestingAncestor(element);
+        ElementInfo ancestorInfo = createInfoForElement(ancestor);
+        ui.Rect globalBounds = getGlobalBounds(element);
+        info = new ElementInfo(ancestorInfo, id, globalBounds);
+        elementToInfoMap[element] = info;
+      }
+      return info;
+    }
+
+    elementToId.keys.forEach(createInfoForElement);
+
+    for (var elementInfo in elementToInfoMap.values) {
+      if (elementInfo.parent == null) {
+        return elementInfo;
+      }
+    }
+
+    return null;
+  }
+
+  static ui.Rect getGlobalBounds(Element element) {
+    RenderObject renderObject = element.renderObject;
+    ui.Rect semanticBounds = renderObject.semanticBounds;
+    Matrix4 globalTransform = renderObject.getTransformTo(null);
+    return MatrixUtils.transformRect(globalTransform, semanticBounds);
+  }
+}
+
+class _DepthFirstElementIterator implements Iterator<Element> {
+  final List<Element> _stack;
+  Element _current;
+
+  _DepthFirstElementIterator(Element rootElement)
+      : _stack = _reverseChildrenOf(rootElement).toList();
+
+  @override
+  Element get current => _current;
+
+  @override
+  bool moveNext() {
+    if (_stack.isEmpty) {
+      return false;
+    }
+
+    _current = _stack.removeLast();
+
+    // Stack children in reverse order to traverse first branch first
+    _stack.addAll(_reverseChildrenOf(_current));
+
+    return true;
+  }
+
+  static Iterable<Element> _reverseChildrenOf(Element element) {
+    assert(element != null);
+    final children = <Element>[];
+    element.debugVisitOnstageChildren(children.add);
+    return children.reversed;
+  }
+}
+
+class _DesignerApp extends StatelessWidget {
+  final Widget widget;
+
+  const _DesignerApp(this.widget);
+
+  @override
+  Widget build(BuildContext context) {
+    return new MaterialApp(
+      title: 'Flutter Demo',
+      theme: new ThemeData(
+        primarySwatch: Colors.blue,
+      ),
+      home: new Scaffold(
+        body: new Center(
+          child: new SizedBox(
+            width: 350.0,
```

Comment:
Yeah, size is an interesting question.

This makes sense it we design a top-level UI.

I will update this eventually to use the size of the preview panel in `Flutter Outline`. Most probably it will be even further from a typical phone, but we will use space more efficiently. Probably OK if we design not a top-level widget.

Ideally we need the whole editor tab space, with ability to resize widget as we want to. Maybe, if we will grow to a full WYSIWYG designer.